aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -297,7 +297,7 @@ impl DelayedFieldRead {
 /// from [SyncCodeCache] it can first check the read-set here.
 enum ModuleRead<DC, VC, S> {
     /// Read from the cross-block module cache.
-    GlobalCache,
+    GlobalCache(Arc<ModuleCode<DC, VC, S>>),
     /// Read from per-block cache ([SyncCodeCache]) used by parallel execution.
     PerBlockCache(Option<(Arc<ModuleCode<DC, VC, S>>, Option<TxnIndex>)>),
```

Comment:
can you explain why do we distinguish reads here based on where we got the data from? also what is Option<TxnIndex> in the PerBlockCache ?

Reply:
Option - module does not exist (in StateView even).

Reply:
Different reads - different validations. We need to check that global reads are still valid, and per-block reads have the same version

Reply:
stupid formatting, didn't show I was referring to TxnIndex

Reply:
Ah - None is a storage version

Reply:
I find it better than `Result<TxnIndex, StorageVersion>`

Reply:
why distinction between storage version and global cache?

Reply:
Different validation paths: for global cache read we need to check if the read is still valid in cache. For per-block we go to MVHashMap. Now, the question is about storage read: we issue it only when there is a cache miss in per-block cache, so it gets validated there.

Reply:
Basically "storage version" can be later drained into global cache, but otherwise exists only in per-block

Reply:
so from validation perspective - there is no distinction

distinction is ONLY there to make updating global cache (i.e. draining to it) be faster/cheaper by skipping things that are already there.

is that correct?

Reply:
actually could be an useful thing to add as a brief comment

Reply:
Added a comment